src/AlarmClock.h

```diff
@@ -39,11 +50,11 @@ template<typename Duration> class AlarmClock {
    }
 
    void Reset() {
-      if (!mExpired.load()) {
-         StopBackgroundThread();
-      }
-      mExpired.store(false); 
-      mExited = std::async(std::launch::async, &AlarmClock::AlarmClockThread, this);
+      unique_lock<mutex> lck(mMutex);
+      mReset.store(true);
+      mExpired.store(0);
```

Comment:
I don't think you need a mutex at all. Only two atomics

/// we can have an assertion that only the owning thread can call this API
// if only one outside thread can call Reset then it simplifies the logic a lot
Reset(){
   mReset(true);
   While(mReset == true){
      Sleep or thread idle
      // waiting for the bg thread to ack the change by setting MReset to false
   }
}